### PR TITLE
core: add support for the "pids" cgroup controller

### PR DESCRIPTION
**Pull request by poettering**
This adds support for the new "pids" cgroup controller of 4.3 kernels.
It allows accounting the number of tasks in a cgroup and enforcing
limits on it.

This adds two new setting TasksAccounting= and TasksMax= to each unit,
as well as a gloabl option DefaultTasksAccounting=.

This also updated "cgtop" to optionally make use of the new
kernel-provided accounting.

systemctl has been updated to show the number of tasks for each service
if it is available.

This patch also adds correct support for undoing memory limits for units
using a MemoryLimit=infinity syntax. We do the same for TasksMax= now
and hence keep things in sync here.

### CODE REVIEW

File: man/systemd.resource-control.xml [hunk 1]
```diff
@@ -197,12 +199,49 @@
...
+          in the same slice and for all its parent slices and the
+          units contained therein. The system default for this setting
+          maybe controlled with
+          <varname>DefaultTasksAccounting=</varname> in
+          <citerefentry><refentrytitle>systemd-system.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry>.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>TasksMax=<replaceable>bytes</replaceable></varname></term>
```
Comment (sztanpet) [inline man/systemd.resource-control.xml]: this is in bytes and the paragraph talks about memory limits but than it talks about the number of processes?

File: man/systemd-cgtop.xml [hunk 2]
```diff
@@ -170,25 +170,48 @@
...
         <term><option>-k</option></term>
 
-        <listitem><para>Include kernel threads when counting tasks in
-        control groups. By default, kernel threads are not included in
-        the count. This setting may also be toggled at runtime by
-        pressing the <keycap>k</keycap> key.</para></listitem>
+        <listitem><para>Count only userspace processes and kernel
+        threads instead of all tasks. By default all tasks are
+        counted: each kernel thread and each userspace thread
+        individually. With this setting kernel threads are included in
```
Comment (sztanpet) [inline man/systemd-cgtop.xml:192]: this sentence is a bit awkward, how about: `By default every kernel and userspace thread is counted.`

also s/maybe/may be/g

Comment (poettering) [inline man/systemd-cgtop.xml:192]: Will fix and force push

File: man/systemd.resource-control.xml (hunk 1 above)
Comment (poettering) [inline man/systemd.resource-control.xml]: will fix an foce push

**Comment by poettering** [comment]
Force pushed.

File: man/systemd-cgtop.xml (hunk 2 above)
Comment (poettering) [inline man/systemd-cgtop.xml:192]: i fixed the "maybe". But I am not sure about your new sentence, as I wanted to put the empasis on _all_ and on "threads individually", both of which got lost in your sentence. I thus didn't change this.

Comment (sztanpet) [inline man/systemd-cgtop.xml:192]: sure, my issue is that the word `individually` is dangling, not a big deal

Thanks!

File: man/systemd.resource-control.xml [hunk 3]
```diff
@@ -197,15 +199,52 @@
...
+
+      <varlistentry>
+        <term><varname>TasksMax=<replaceable>N</replaceable></varname></term>
+
+        <listitem>
+          <para>Specify the maximum number of tasks that may be
+          created in the unit. This ensures that the number of tasks
+          accounted for the unit (see above) stays below a specific
+          limit. If assigned the special value
+          <literal>infinity</literal> no memory limit is applied. This
```
Comment (sztanpet) [inline man/systemd.resource-control.xml]: still talks about memory limit though :)

**Comment by poettering** [comment]
@sztanpet fixed the memory limit thing now too. thanks!

**Pull request merged**